homeassistant/components/enasolar/sensor.py

```diff
@@ -0,0 +1,288 @@
+
+    enasolar.setup_sensors()
+
+    for sensor in enasolar.sensors:
+        _LOGGER.debug("Setup sensor %s", sensor.key)
+        if sensor.enabled:
+            if sensor.is_meter:
+                hass_meter_sensors.append(
+                    EnaSolarSensor(sensor, inverter_name=config_entry.data[CONF_NAME])
+                )
+            else:
+                hass_data_sensors.append(
+                    EnaSolarSensor(sensor, inverter_name=config_entry.data[CONF_NAME])
+                )
+
+    async_add_entities(hass_meter_sensors)
+    async_add_entities(hass_data_sensors)
+
+    async def async_enasolar_meters():
+        """Update the EnaSolar Meter sensors."""
+
+        if enasolar.sun_up <= datetime.now().time() <= enasolar.sun_down:
+            values = await enasolar.read_meters()
+        else:
+            values = False
+
+        for sensor in hass_meter_sensors:
+            state_unknown = False
+            if not values and (
+                (sensor.per_day_basis and date.today() > sensor.date_updated)
+                or (not sensor.per_day_basis and not sensor.per_total_basis)
+            ):
+                state_unknown = True
+            sensor.async_update_values(unknown_state=state_unknown)
+            _LOGGER.debug(
+                "Meter Sensor %s updated => %s", sensor._sensor.key, sensor.state
+            )
+
+        return values
+
+    async def async_enasolar_data():
+        """Update the EnaSolar Data sensors."""
+
+        if enasolar.sun_up <= datetime.now().time() <= enasolar.sun_down:
+            values = await enasolar.read_data()
+        else:
+            values = False
+
+        for sensor in hass_data_sensors:
+            state_unknown = False
+            if not values and (
+                (sensor.per_day_basis and date.today() > sensor.date_updated)
+                or (not sensor.per_day_basis and not sensor.per_total_basis)
+            ):
+                state_unknown = True
+            sensor.async_update_values(unknown_state=state_unknown)
+            _LOGGER.debug(
+                "Data Sensor %s updated => %s", sensor._sensor.key, sensor.state
+            )
+
+        return values
+
+    def start_update_interval(event):
+        """Start the update interval scheduling."""
+        nonlocal remove_interval_update_meters, remove_interval_update_data
+
+        remove_interval_update_meters = async_track_time_interval_backoff(
+            hass, async_enasolar_meters, SCAN_METERS_MIN_INTERVAL
+        )
+        remove_interval_update_data = async_track_time_interval_backoff(
+            hass, async_enasolar_data, SCAN_DATA_MIN_INTERVAL
+        )
+
+    def stop_update_interval(event):
+        """Properly cancel the scheduled update."""
+        remove_interval_update_meters()  # pylint: disable=not-callable
+        remove_interval_update_data()  # pylint: disable=not-callable
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_update_interval)
```

Comment:
I think you'd be able to simplify all of this update logic significantly if you moved to using a `DataUpdateCoordinator`. Your update logic and your interval backoff logic could live in there with much less code and then you could inherit from `CoordinatorEntity` in your sensor class so that your sensors get signaled by the coordinator on each successful update

Reply:
You seem to be reaching into the entity class instances as part of your update. You can store things in `hass.data[DOMAIN]` like your sensors so they are accessible in the coordinator

Reply:
For now I'd like to keep the login as it is, if I start feeling very adventurous I might do this in a later version